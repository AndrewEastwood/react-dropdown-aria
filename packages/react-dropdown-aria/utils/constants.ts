export enum StyleKeys {
  Arrow = 'Arrow',
  DropdownSelector = 'DropdownSelector',
  DropdownWrapper = 'DropdownWrapper',
  SelectorSearch = 'SelectorSearch',
  SelectedValue = 'SelectedValue',
  Placeholder = 'Placeholder',
  GroupContainer = 'GroupContainer',
  GroupDivider = 'GroupDivider',
  GroupHeading = 'GroupHeading',
  OptionContainer = 'OptionContainer',
  OptionItem = 'OptionItem',
}

export enum KEY_CODES {
  DOWN_ARROW = 40,
  ENTER = 13,
  ESCAPE = 27,
  PAGE_DOWN = 34,
  PAGE_UP = 33,
  TAB = 9,
  UP_ARROW = 38,
}

export const NAVIGATION_KEYS = [
  KEY_CODES.ESCAPE,
  KEY_CODES.UP_ARROW,
  KEY_CODES.DOWN_ARROW,
  KEY_CODES.PAGE_UP,
  KEY_CODES.PAGE_DOWN,
];

export const IdPrefix = 'react_dropdown_aria_';
