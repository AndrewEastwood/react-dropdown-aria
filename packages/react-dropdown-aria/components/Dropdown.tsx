import React, { KeyboardEvent, useCallback, ChangeEvent, useMemo } from 'react';
import { KEY_CODES, NAVIGATION_KEYS } from '../utils/constants';
import DropdownContent from './DropdownContent';
import { DropdownProps } from '../utils/types';
import { cx } from '../utils/helper';
import { useDropdownHooks, useId } from '../utils/dropdown-hooks';
import { ChevronDown, Search } from './icons';

import {
  DropdownWrapper,
  DropdownSelector,
  SelectorSearch,
  SelectedValue,
  Placeholder,
  Arrow,
  OptionContainer,
} from '../styles';

const Dropdown = (props: DropdownProps) => {
  const {
    arrowRenderer,
    contentClassName,
    className,
    disabled,
    hideArrow,
    id,
    optionItemRenderer,
    pageKeyTraverseSize,
    placeholder,
    searchable,
    value,
    selectedValueClassName,
  } = props;

  const mergedId = useId(id);

  const {
    open,
    dropdownFocused,
    focusedIndex,
    setFocusedIndex,
    setDropdownFocused,
    setValue,
    openDropdown,
    closeDropdown,
    searchTerm,
    setSearchTerm,
    filteredOptions,
    flattenedOptions,
    container,
    inputRef,
    listWrapper,
    ariaProps,
    ariaList,
  } = useDropdownHooks(props, mergedId);

  // Pass focus on to input
  const forwardFocus = useCallback(() => {
    if (inputRef.current) {
      inputRef.current.focus();
    }
  }, [inputRef.current]);

  const onDropdownClick = useCallback(() => {
    forwardFocus();
    if (!disabled && (!open || !searchable)) {
      if (open) {
        closeDropdown(true);
      } else {
        openDropdown();
      }
    }
  }, [open, disabled, searchable, closeDropdown, openDropdown]);

  const onNavigation = useCallback(
    (keyCode: number) => {
      switch (keyCode) {
        case KEY_CODES.UP_ARROW:
          setFocusedIndex(prev => {
            if (prev === 0) return flattenedOptions.length - 1;
            return prev - 1;
          });
          break;
        case KEY_CODES.DOWN_ARROW:
          setFocusedIndex(p => (p + 1) % flattenedOptions.length);
          break;
        case KEY_CODES.PAGE_UP:
          setFocusedIndex(prev => {
            if (prev - pageKeyTraverseSize < 0 && prev !== 0) return 0;
            if (prev - pageKeyTraverseSize < 0) return flattenedOptions.length - 1;
            return prev - pageKeyTraverseSize;
          });
          break;
        case KEY_CODES.PAGE_DOWN:
          setFocusedIndex(prev => {
            if (prev === flattenedOptions.length - 1) return 0;
            if (prev + pageKeyTraverseSize > flattenedOptions.length - 1) return flattenedOptions.length - 1;
            return (prev + pageKeyTraverseSize) % flattenedOptions.length;
          });
          break;
        case KEY_CODES.ESCAPE:
          closeDropdown(true);
          break;
        default:
          break;
      }
    },
    [setFocusedIndex, flattenedOptions, pageKeyTraverseSize, closeDropdown],
  );

  const handleInputKeyDown = useCallback(
    (e: KeyboardEvent) => {
      const { keyCode } = e;

      if (NAVIGATION_KEYS.indexOf(keyCode) !== -1) {
        e.preventDefault();
        e.stopPropagation();
        onNavigation(keyCode);
      } else if ((keyCode === KEY_CODES.ENTER || (keyCode === KEY_CODES.SPACE && !searchable)) && !open) {
        e.preventDefault();
        openDropdown();
      } else if (keyCode === KEY_CODES.TAB && !searchable) {
        closeDropdown();
      } else if (
        (keyCode === KEY_CODES.TAB || keyCode === KEY_CODES.ENTER) &&
        flattenedOptions.length > 0 &&
        focusedIndex >= 0 &&
        open
      ) {
        e.stopPropagation();
        e.preventDefault();
        setValue(flattenedOptions[focusedIndex], true);
      }
    },
    [flattenedOptions, setValue, focusedIndex, open, onNavigation, openDropdown, searchable, closeDropdown],
  );

  const handleTermChange = useCallback(
    (e: ChangeEvent<HTMLInputElement>) => {
      setSearchTerm(e.target.value);
    },
    [setSearchTerm],
  );

  const onFocus = useCallback(() => setDropdownFocused(true), [setDropdownFocused]);
  const onBlur = useCallback(() => setDropdownFocused(false), [setDropdownFocused]);

  // ---------------------- RENDER -----------------------

  const ArrowMarkup = useMemo(() => {
    if (hideArrow) return null;
    if (arrowRenderer) return <Arrow className={'dropdown-arrow'}>{arrowRenderer(open)}</Arrow>;

    const showSearchIcon = open && searchable;
    return (
      <Arrow className={'dropdown-arrow'}>
        {showSearchIcon && <Search />}
        {!showSearchIcon && <ChevronDown />}
      </Arrow>
    );
  }, [open, arrowRenderer, searchable, hideArrow]);

  return (
    <DropdownWrapper
      ref={container}
      onFocus={forwardFocus}
      onClick={onDropdownClick}
      role="button"
      width={props.width}
      height={props.height}
      disabled={disabled}
      open={open}
      dropdownFocused={dropdownFocused}
      className={cx('dropdown', className)}
    >
      <DropdownSelector className={'dropdown-selector'} open={open} searchable={props.searchable}>
        <SelectorSearch className={'dropdown-selector-search'}>
          <input
            id={mergedId}
            ref={inputRef}
            value={searchTerm}
            onChange={handleTermChange}
            onKeyDown={handleInputKeyDown}
            onFocus={onFocus}
            onBlur={onBlur}
            readOnly={!open || !searchable}
            disabled={disabled}
            autoComplete="off"
            role="combobox"
            {...ariaProps}
          />
        </SelectorSearch>
        {!value && !searchTerm && (
          <Placeholder className={'dropdown-selector-placeholder'} centerText={props.centerText}>
            {placeholder}
          </Placeholder>
        )}
        {value && !searchTerm && (
          <SelectedValue
            className={cx('dropdown-selector-value', selectedValueClassName)}
            centerText={props.centerText}
            value={value}
            open={open}
          >
            {value}
          </SelectedValue>
        )}
        {ArrowMarkup}
      </DropdownSelector>
      {ariaList}
      <OptionContainer
        maxContentHeight={props.maxContentHeight}
        openUp={props.openUp}
        open={open}
        className={cx('dropdown-selector-content', contentClassName)}
        ref={listWrapper}
      >
        <DropdownContent
          selectedOption={value}
          options={filteredOptions}
          focusedIndex={focusedIndex}
          onOptionClicked={setValue}
          optionItemRenderer={optionItemRenderer}
          empty={flattenedOptions.length === 0}
        />
      </OptionContainer>
    </DropdownWrapper>
  );
};

Dropdown.defaultProps = {
  ariaDescribedBy: null,
  ariaLabel: null,
  ariaLabelledBy: null,
  arrowRenderer: undefined,
  centerText: false,
  className: undefined,
  contentClassName: null,
  defaultOpen: false,
  disabled: false,
  height: null,
  hideArrow: false,
  id: null,
  maxContentHeight: 150,
  openUp: false,
  optionItemRenderer: undefined,
  pageKeyTraverseSize: 10,
  placeholder: 'Select ...',
  searchable: false,
  selectedValueClassName: null,
  style: {},
  value: undefined,
  width: null,
};

export default Dropdown;
