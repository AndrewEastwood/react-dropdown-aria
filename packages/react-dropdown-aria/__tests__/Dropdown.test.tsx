import React from 'react';
import { mount, ReactWrapper, HTMLAttributes } from 'enzyme';
import sinon from 'sinon';
import toJson from 'enzyme-to-json';
import Dropdown, { Option } from '../index';
import DropdownTest from './DropdownTest';
import { CUSTOM_OPTIONS, OPTIONS, GROUPED_OPTIONS } from './constants';
import { KEY_CODES } from '../utils/constants';

// eslint-disable-next-line @typescript-eslint/no-empty-function, @typescript-eslint/no-unused-vars
const foo = (val: Option) => {};

describe('Check Props', () => {
  it('Matches snapshot with default props', () => {
    const dropdown = mount(<Dropdown onChange={foo} options={OPTIONS} />);
    expect(toJson(dropdown)).toMatchSnapshot();
  });

  it('Matches snapshot with custom props', () => {
    const dropdown = mount(
      <Dropdown
        placeholder="Custom Placeholder..."
        className="custom-class"
        id="dropdown"
        ariaLabel="React Simple Dropdown"
        options={CUSTOM_OPTIONS}
        onChange={foo}
        disabled
        width={400}
        maxContentHeight={150}
      />,
    );
    expect(toJson(dropdown)).toMatchSnapshot();
  });

  it('Matches snapshot with grouped options', () => {
    const dropdown = mount(<Dropdown onChange={foo} options={GROUPED_OPTIONS} />);
    expect(toJson(dropdown)).toMatchSnapshot();
  });
});

describe('Navigation', () => {
  let wrapper: ReactWrapper<any, Readonly<{}>, React.Component<{}, {}, any>>;
  let button: ReactWrapper<any, Readonly<{}>, React.Component<{}, {}, any>>;
  let input: ReactWrapper<any, Readonly<{}>, React.Component<{}, {}, any>>;
  let listContainer: ReactWrapper<HTMLAttributes, Readonly<{}>, React.Component<{}, {}, any>>;

  beforeEach(() => {
    wrapper = mount(<DropdownTest onChange={foo} />);
    button = wrapper.find('.test').first(); // Actual dropdown button element
    listContainer = wrapper.find('.dropdown-selector-content').first();
    input = wrapper.find('input').first();


    // Do some action so that focus is in correct place
    wrapper.simulate('keyDown', { key: 'downArrow', keyCode: KEY_CODES.DOWN_ARROW, preventDefault: foo });
  });

  const dropdownValue = () => wrapper.find('.dropdown-selector-value').first().text();
  const getWrapperDisplayProp = () => getComputedStyle(listContainer.getDOMNode()).getPropertyValue('display');
  const pressEnter = (comp = input) =>  comp.simulate('keyDown', { key: 'enter', keyCode: KEY_CODES.ENTER, preventDefault: foo });
  const keyDown = (comp = input) => comp.simulate('keyDown', { key: 'downArrow', keyCode: KEY_CODES.DOWN_ARROW, preventDefault: foo });

  it('Opens dropdown when clicked', () => {
    button.simulate('click');

    expect(getWrapperDisplayProp()).toBe('block');
  });

  it('Opens dropdown with enter key', () => {
    input.simulate('keyDown', { key: 'enter', keyCode: KEY_CODES.ENTER, preventDefault: foo });

    expect(getWrapperDisplayProp()).toBe('block');
  });

  it('Closes when clicked again', () => {
    button.simulate('click');
    expect(getWrapperDisplayProp()).toBe('block');

    button.simulate('click');
    expect(getWrapperDisplayProp()).toBe('none');
  });

  it('Closes when enter pressed again', () => {
    pressEnter(input)
    expect(getWrapperDisplayProp()).toBe('block');

    pressEnter(input)
    expect(getWrapperDisplayProp()).toBe('none');
  });

  it('Closes when tab key pressed', () => {
    button.simulate('click');
    expect(getWrapperDisplayProp()).toBe('block');

    input.simulate('keyDown', { key: 'tab', keyCode: KEY_CODES.TAB });
    expect(getWrapperDisplayProp()).toBe('none');
  });

  it('Closes when esc pressed', () => {
    button.simulate('click');
    expect(getWrapperDisplayProp()).toBe('block');

    input.simulate('keyDown', { key: 'escape', keyCode: KEY_CODES.ESCAPE, preventDefault: foo });
    expect(getWrapperDisplayProp()).toBe('none');
  });

  it('Arrow key selects first element in list', () => {
    input.simulate('click');
    pressEnter(input);
    pressEnter(input);

    expect(dropdownValue()).toBe('1');
  });

  it('Arrow key nav loops arround', () => {
    button.simulate('click');
    for (let i = 0; i < 10; i += 1) {
      keyDown(input);
    }
    pressEnter(input);

    expect(dropdownValue()).toBe('1');
  });
});

describe('Selecting Options', () => {
  let spy: any;

  beforeEach(() => {
    spy = sinon.spy();
  });

  it('calls onChange with value when option selected', () => {
    const wrapper = mount(<Dropdown options={OPTIONS} onChange={spy} className="test" />);
    const button = wrapper.find('.test').first();

    button.simulate('click');
    wrapper.find('.dropdown-option').first().simulate('click');

    expect(spy.calledWith({ value: '1' })).toBe(true);
  });

  it('sets dropdown text to selected item', () => {
    const wrapper = mount(<DropdownTest onChange={foo} />);
    const button = wrapper.find('.test').first();

    button.simulate('click');
    button.find('.dropdown-option').first().simulate('click');

    expect(wrapper.find('.dropdown-selector-value').first().text()).toBe('1');
  });
});

describe('Special props', () => {
  it('Does not open when disabled', () => {
    const wrapper = mount(<Dropdown options={OPTIONS} onChange={foo} className="test" disabled />);
    wrapper.find('.test').first().simulate('click');
    const listContainer = wrapper.find('.dropdown-selector-content').first();

    expect(getComputedStyle(listContainer.getDOMNode()).getPropertyValue('display')).toBe('none');
  });

  it('Does not open by space when searchable enabled', () => {
    const wrapper = mount(<Dropdown options={OPTIONS} onChange={foo} searchable />);
    wrapper.find('input').first().simulate('keyDown', { key: 'space', keyCode: KEY_CODES.SPACE, preventDefault: foo });
    const listContainer = wrapper.find('.dropdown-selector-content').first();

    expect(getComputedStyle(listContainer.getDOMNode()).getPropertyValue('display')).toBe('none');
  });

  it('Opens by space when searchable disabled', () => {
    const wrapper = mount(<Dropdown options={OPTIONS} onChange={foo} searchable={false} />);
    wrapper.find('input').first().simulate('keyDown', { key: 'space', keyCode: KEY_CODES.SPACE, preventDefault: foo });
    const listContainer = wrapper.find('.dropdown-selector-content').first();

    expect(getComputedStyle(listContainer.getDOMNode()).getPropertyValue('display')).toBe('block');
  });
});
