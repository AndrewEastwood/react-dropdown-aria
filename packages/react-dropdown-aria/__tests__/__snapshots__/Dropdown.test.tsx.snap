// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check Props Matches snapshot with custom props 1`] = `
<Dropdown
  ariaDescribedBy={null}
  ariaLabel="React Simple Dropdown"
  ariaLabelledBy={null}
  centerText={false}
  className="custom-class"
  contentClassName={null}
  disabled={true}
  height={null}
  hideArrow={false}
  id="dropdown"
  maxContentHeight={150}
  onChange={[Function]}
  openUp={false}
  options={
    Array [
      Object {
        "ariaLabel": "ariaLabel",
        "className": "className",
        "iconClass": "fab fa-js-square",
        "title": "title",
        "value": "1",
      },
      Object {
        "ariaLabel": "ariaLabel",
        "className": "className",
        "iconClass": "fab fa-js-square",
        "title": "title",
        "value": "2",
      },
      Object {
        "ariaLabel": "ariaLabel",
        "className": "className",
        "iconClass": "fab fa-js-square",
        "title": "title",
        "value": "3",
      },
    ]
  }
  pageKeyTraverseSize={10}
  placeholder="Custom Placeholder..."
  searchable={false}
  selectedValueClassName={null}
  style={Object {}}
  width={400}
>
  <div
    aria-describedby={null}
    aria-hidden={true}
    aria-label="React Simple Dropdown"
    aria-labelledby={null}
    className="dropdown custom-class css-w4jram"
    id="dropdown"
    onClick={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
  >
    <div
      className="dropdown-selector css-q0xgr4"
    >
      <span
        className="dropdown-selector-search css-1l6e6we"
      >
        <input
          autoComplete="off"
          disabled={true}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          readOnly={true}
          value=""
        />
      </span>
      <span
        className="dropdown-selector-placeholder css-nctil"
      >
        Custom Placeholder...
      </span>
      <Arrow
        dropdownOpen={false}
        getStyle={[Function]}
        hideArrow={false}
        searchable={false}
      >
        <div
          className="css-1h2654e"
        >
          <ChevronDown>
            <svg
              className="feather feather-chevron-down"
              fill="none"
              height="16"
              stroke="currentColor"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth="2"
              viewBox="0 0 24 24"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <polyline
                points="6 9 12 15 18 9"
              />
            </svg>
          </ChevronDown>
        </div>
      </Arrow>
    </div>
    <ul
      className="dropdown-selector-content css-e99cqv"
    >
      <OptionItem
        key="1"
        onOptionClicked={[Function]}
        option={
          Object {
            "ariaLabel": "ariaLabel",
            "className": "className",
            "iconClass": "fab fa-js-square",
            "title": "title",
            "value": "1",
          }
        }
        optionClass="className css-qte9rq"
      >
        <div
          aria-label="ariaLabel"
          className="dropdown-option className css-qte9rq"
          onClick={[Function]}
          tabIndex={-1}
          title="title"
        >
          <i
            className="fab fa-js-square dropdown-option-icon"
          />
          1
        </div>
      </OptionItem>
      <OptionItem
        key="2"
        onOptionClicked={[Function]}
        option={
          Object {
            "ariaLabel": "ariaLabel",
            "className": "className",
            "iconClass": "fab fa-js-square",
            "title": "title",
            "value": "2",
          }
        }
        optionClass="className css-qte9rq"
      >
        <div
          aria-label="ariaLabel"
          className="dropdown-option className css-qte9rq"
          onClick={[Function]}
          tabIndex={-1}
          title="title"
        >
          <i
            className="fab fa-js-square dropdown-option-icon"
          />
          2
        </div>
      </OptionItem>
      <OptionItem
        key="3"
        onOptionClicked={[Function]}
        option={
          Object {
            "ariaLabel": "ariaLabel",
            "className": "className",
            "iconClass": "fab fa-js-square",
            "title": "title",
            "value": "3",
          }
        }
        optionClass="className css-qte9rq"
      >
        <div
          aria-label="ariaLabel"
          className="dropdown-option className css-qte9rq"
          onClick={[Function]}
          tabIndex={-1}
          title="title"
        >
          <i
            className="fab fa-js-square dropdown-option-icon"
          />
          3
        </div>
      </OptionItem>
    </ul>
  </div>
</Dropdown>
`;

exports[`Check Props Matches snapshot with default props 1`] = `
<Dropdown
  ariaDescribedBy={null}
  ariaLabel={null}
  ariaLabelledBy={null}
  centerText={false}
  contentClassName={null}
  disabled={false}
  height={null}
  hideArrow={false}
  id={null}
  maxContentHeight={150}
  onChange={[Function]}
  openUp={false}
  options={Array []}
  pageKeyTraverseSize={10}
  placeholder="Select ..."
  searchable={false}
  selectedValueClassName={null}
  style={Object {}}
  width={null}
>
  <div
    aria-describedby={null}
    aria-hidden={false}
    aria-label={null}
    aria-labelledby={null}
    className="dropdown css-4a6799"
    id={null}
    onClick={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
  >
    <div
      className="dropdown-selector css-q0xgr4"
    >
      <span
        className="dropdown-selector-search css-1l6e6we"
      >
        <input
          autoComplete="off"
          disabled={false}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          readOnly={true}
          value=""
        />
      </span>
      <span
        className="dropdown-selector-placeholder css-nctil"
      >
        Select ...
      </span>
      <Arrow
        dropdownOpen={false}
        getStyle={[Function]}
        hideArrow={false}
        searchable={false}
      >
        <div
          className="css-1h2654e"
        >
          <ChevronDown>
            <svg
              className="feather feather-chevron-down"
              fill="none"
              height="16"
              stroke="currentColor"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth="2"
              viewBox="0 0 24 24"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <polyline
                points="6 9 12 15 18 9"
              />
            </svg>
          </ChevronDown>
        </div>
      </Arrow>
    </div>
    <ul
      className="dropdown-selector-content css-e99cqv"
    >
      <div
        className="dropdown-selector-content--empty"
      >
        <Inbox>
          <svg
            className="feather feather-chevron-down"
            fill="none"
            height="24"
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <polyline
              points="22 12 16 12 14 15 10 15 8 12 2 12"
            />
            <path
              d="M5.45 5.11L2 12v6a2 2 0 0 0 2 2h16a2 2 0 0 0 2-2v-6l-3.45-6.89A2 2 0 0 0 16.76 4H7.24a2 2 0 0 0-1.79 1.11z"
            />
          </svg>
        </Inbox>
        No data
      </div>
    </ul>
  </div>
</Dropdown>
`;

exports[`Check Props Matches snapshot with grouped options 1`] = `
<Dropdown
  ariaDescribedBy={null}
  ariaLabel={null}
  ariaLabelledBy={null}
  centerText={false}
  contentClassName={null}
  disabled={false}
  height={null}
  hideArrow={false}
  id={null}
  maxContentHeight={150}
  onChange={[Function]}
  openUp={false}
  options={
    Array [
      Object {
        "groupOptions": Array [
          Object {
            "value": "1",
          },
          Object {
            "value": "2",
          },
          Object {
            "value": "3",
          },
          Object {
            "value": "4",
          },
          Object {
            "value": "5",
          },
          Object {
            "value": "6",
          },
          Object {
            "value": "7",
          },
          Object {
            "value": "8",
          },
          Object {
            "value": "9",
          },
          Object {
            "value": "10",
          },
        ],
        "label": "numbers",
      },
      Object {
        "groupOptions": Array [
          Object {
            "value": "a",
          },
          Object {
            "value": "b",
          },
          Object {
            "value": "c",
          },
          Object {
            "value": "d",
          },
          Object {
            "value": "e",
          },
          Object {
            "value": "f",
          },
          Object {
            "value": "g",
          },
          Object {
            "value": "h",
          },
          Object {
            "value": "i",
          },
          Object {
            "value": "j",
          },
        ],
        "label": "letters",
      },
    ]
  }
  pageKeyTraverseSize={10}
  placeholder="Select ..."
  searchable={false}
  selectedValueClassName={null}
  style={Object {}}
  width={null}
>
  <div
    aria-describedby={null}
    aria-hidden={false}
    aria-label={null}
    aria-labelledby={null}
    className="dropdown css-4a6799"
    id={null}
    onClick={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
  >
    <div
      className="dropdown-selector css-q0xgr4"
    >
      <span
        className="dropdown-selector-search css-1l6e6we"
      >
        <input
          autoComplete="off"
          disabled={false}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          readOnly={true}
          value=""
        />
      </span>
      <span
        className="dropdown-selector-placeholder css-nctil"
      >
        Select ...
      </span>
      <Arrow
        dropdownOpen={false}
        getStyle={[Function]}
        hideArrow={false}
        searchable={false}
      >
        <div
          className="css-1h2654e"
        >
          <ChevronDown>
            <svg
              className="feather feather-chevron-down"
              fill="none"
              height="16"
              stroke="currentColor"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth="2"
              viewBox="0 0 24 24"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <polyline
                points="6 9 12 15 18 9"
              />
            </svg>
          </ChevronDown>
        </div>
      </Arrow>
    </div>
    <ul
      className="dropdown-selector-content css-e99cqv"
    >
      <OptionGroup
        focusedIndex={-1}
        getStyle={[Function]}
        key="numbers"
        onOptionClicked={[Function]}
        optionGroup={
          Object {
            "groupOptions": Array [
              Object {
                "value": "1",
              },
              Object {
                "value": "2",
              },
              Object {
                "value": "3",
              },
              Object {
                "value": "4",
              },
              Object {
                "value": "5",
              },
              Object {
                "value": "6",
              },
              Object {
                "value": "7",
              },
              Object {
                "value": "8",
              },
              Object {
                "value": "9",
              },
              Object {
                "value": "10",
              },
            ],
            "label": "numbers",
          }
        }
        startingIndex={0}
      >
        <div
          className="css-dckqxd"
        >
          <div
            className="css-1q3omhu"
          >
            <div>
              NUMBERS
               |  
            </div>
            <div>
              10
            </div>
          </div>
          <OptionItem
            key="1"
            onOptionClicked={[Function]}
            option={
              Object {
                "value": "1",
              }
            }
            optionClass="css-qte9rq"
          >
            <div
              className="dropdown-option css-qte9rq"
              onClick={[Function]}
              tabIndex={-1}
            >
              1
            </div>
          </OptionItem>
          <OptionItem
            key="2"
            onOptionClicked={[Function]}
            option={
              Object {
                "value": "2",
              }
            }
            optionClass="css-qte9rq"
          >
            <div
              className="dropdown-option css-qte9rq"
              onClick={[Function]}
              tabIndex={-1}
            >
              2
            </div>
          </OptionItem>
          <OptionItem
            key="3"
            onOptionClicked={[Function]}
            option={
              Object {
                "value": "3",
              }
            }
            optionClass="css-qte9rq"
          >
            <div
              className="dropdown-option css-qte9rq"
              onClick={[Function]}
              tabIndex={-1}
            >
              3
            </div>
          </OptionItem>
          <OptionItem
            key="4"
            onOptionClicked={[Function]}
            option={
              Object {
                "value": "4",
              }
            }
            optionClass="css-qte9rq"
          >
            <div
              className="dropdown-option css-qte9rq"
              onClick={[Function]}
              tabIndex={-1}
            >
              4
            </div>
          </OptionItem>
          <OptionItem
            key="5"
            onOptionClicked={[Function]}
            option={
              Object {
                "value": "5",
              }
            }
            optionClass="css-qte9rq"
          >
            <div
              className="dropdown-option css-qte9rq"
              onClick={[Function]}
              tabIndex={-1}
            >
              5
            </div>
          </OptionItem>
          <OptionItem
            key="6"
            onOptionClicked={[Function]}
            option={
              Object {
                "value": "6",
              }
            }
            optionClass="css-qte9rq"
          >
            <div
              className="dropdown-option css-qte9rq"
              onClick={[Function]}
              tabIndex={-1}
            >
              6
            </div>
          </OptionItem>
          <OptionItem
            key="7"
            onOptionClicked={[Function]}
            option={
              Object {
                "value": "7",
              }
            }
            optionClass="css-qte9rq"
          >
            <div
              className="dropdown-option css-qte9rq"
              onClick={[Function]}
              tabIndex={-1}
            >
              7
            </div>
          </OptionItem>
          <OptionItem
            key="8"
            onOptionClicked={[Function]}
            option={
              Object {
                "value": "8",
              }
            }
            optionClass="css-qte9rq"
          >
            <div
              className="dropdown-option css-qte9rq"
              onClick={[Function]}
              tabIndex={-1}
            >
              8
            </div>
          </OptionItem>
          <OptionItem
            key="9"
            onOptionClicked={[Function]}
            option={
              Object {
                "value": "9",
              }
            }
            optionClass="css-qte9rq"
          >
            <div
              className="dropdown-option css-qte9rq"
              onClick={[Function]}
              tabIndex={-1}
            >
              9
            </div>
          </OptionItem>
          <OptionItem
            key="10"
            onOptionClicked={[Function]}
            option={
              Object {
                "value": "10",
              }
            }
            optionClass="css-qte9rq"
          >
            <div
              className="dropdown-option css-qte9rq"
              onClick={[Function]}
              tabIndex={-1}
            >
              10
            </div>
          </OptionItem>
          <div
            className="css-iryzig"
          />
        </div>
      </OptionGroup>
      <OptionGroup
        focusedIndex={-1}
        getStyle={[Function]}
        key="letters"
        onOptionClicked={[Function]}
        optionGroup={
          Object {
            "groupOptions": Array [
              Object {
                "value": "a",
              },
              Object {
                "value": "b",
              },
              Object {
                "value": "c",
              },
              Object {
                "value": "d",
              },
              Object {
                "value": "e",
              },
              Object {
                "value": "f",
              },
              Object {
                "value": "g",
              },
              Object {
                "value": "h",
              },
              Object {
                "value": "i",
              },
              Object {
                "value": "j",
              },
            ],
            "label": "letters",
          }
        }
        startingIndex={10}
      >
        <div
          className="css-dckqxd"
        >
          <div
            className="css-1q3omhu"
          >
            <div>
              LETTERS
               |  
            </div>
            <div>
              10
            </div>
          </div>
          <OptionItem
            key="a"
            onOptionClicked={[Function]}
            option={
              Object {
                "value": "a",
              }
            }
            optionClass="css-qte9rq"
          >
            <div
              className="dropdown-option css-qte9rq"
              onClick={[Function]}
              tabIndex={-1}
            >
              a
            </div>
          </OptionItem>
          <OptionItem
            key="b"
            onOptionClicked={[Function]}
            option={
              Object {
                "value": "b",
              }
            }
            optionClass="css-qte9rq"
          >
            <div
              className="dropdown-option css-qte9rq"
              onClick={[Function]}
              tabIndex={-1}
            >
              b
            </div>
          </OptionItem>
          <OptionItem
            key="c"
            onOptionClicked={[Function]}
            option={
              Object {
                "value": "c",
              }
            }
            optionClass="css-qte9rq"
          >
            <div
              className="dropdown-option css-qte9rq"
              onClick={[Function]}
              tabIndex={-1}
            >
              c
            </div>
          </OptionItem>
          <OptionItem
            key="d"
            onOptionClicked={[Function]}
            option={
              Object {
                "value": "d",
              }
            }
            optionClass="css-qte9rq"
          >
            <div
              className="dropdown-option css-qte9rq"
              onClick={[Function]}
              tabIndex={-1}
            >
              d
            </div>
          </OptionItem>
          <OptionItem
            key="e"
            onOptionClicked={[Function]}
            option={
              Object {
                "value": "e",
              }
            }
            optionClass="css-qte9rq"
          >
            <div
              className="dropdown-option css-qte9rq"
              onClick={[Function]}
              tabIndex={-1}
            >
              e
            </div>
          </OptionItem>
          <OptionItem
            key="f"
            onOptionClicked={[Function]}
            option={
              Object {
                "value": "f",
              }
            }
            optionClass="css-qte9rq"
          >
            <div
              className="dropdown-option css-qte9rq"
              onClick={[Function]}
              tabIndex={-1}
            >
              f
            </div>
          </OptionItem>
          <OptionItem
            key="g"
            onOptionClicked={[Function]}
            option={
              Object {
                "value": "g",
              }
            }
            optionClass="css-qte9rq"
          >
            <div
              className="dropdown-option css-qte9rq"
              onClick={[Function]}
              tabIndex={-1}
            >
              g
            </div>
          </OptionItem>
          <OptionItem
            key="h"
            onOptionClicked={[Function]}
            option={
              Object {
                "value": "h",
              }
            }
            optionClass="css-qte9rq"
          >
            <div
              className="dropdown-option css-qte9rq"
              onClick={[Function]}
              tabIndex={-1}
            >
              h
            </div>
          </OptionItem>
          <OptionItem
            key="i"
            onOptionClicked={[Function]}
            option={
              Object {
                "value": "i",
              }
            }
            optionClass="css-qte9rq"
          >
            <div
              className="dropdown-option css-qte9rq"
              onClick={[Function]}
              tabIndex={-1}
            >
              i
            </div>
          </OptionItem>
          <OptionItem
            key="j"
            onOptionClicked={[Function]}
            option={
              Object {
                "value": "j",
              }
            }
            optionClass="css-qte9rq"
          >
            <div
              className="dropdown-option css-qte9rq"
              onClick={[Function]}
              tabIndex={-1}
            >
              j
            </div>
          </OptionItem>
          <div
            className="css-iryzig"
          />
        </div>
      </OptionGroup>
    </ul>
  </div>
</Dropdown>
`;
